File: test-qso.py
#!/usr/bin/env python3
"""
Test QSO execution via MCP web API
Completely independent from the MCP server - uses HTTP API only
"""

import requests
import json
import time
import sys

MCP_API_URL = "http://localhost:3001/api"

def get_slices():
    """Get current slice states with decoded stations"""
    response = requests.get(f"{MCP_API_URL}/slices")
    response.raise_for_status()
    return response.json()

def find_cq_stations(slices, channel='C', min_snr=-20, max_snr=-10):
    """Find stations calling CQ on specified channel within SNR range"""
    cq_stations = []

    for slice_data in slices:
        # API uses 'id' not 'channel'
        if slice_data.get('id') == channel or slice_data.get('name') == channel:
            stations = slice_data.get('stations', [])
            for station in stations:
                message = station.get('message', '')
                snr = station.get('snr', 0)
                callsign = station.get('callsign', '')
                grid = station.get('grid', '')

                # Look for CQ messages in SNR range
                if 'CQ' in message and min_snr <= snr <= max_snr and callsign:
                    cq_stations.append({
                        'callsign': callsign,
                        'grid': grid,
                        'snr': snr,
                        'message': message,
                        'time': station.get('lastSeen', '')
                    })

    return cq_stations

def execute_qso(instance_id, target_callsign, my_callsign, my_grid):
    """Execute QSO via web API"""
    payload = {
        'instanceId': instance_id,
        'targetCallsign': target_callsign,
        'myCallsign': my_callsign,
        'myGrid': my_grid
    }

    response = requests.post(f"{MCP_API_URL}/qso/execute", json=payload)
    response.raise_for_status()
    return response.json()

def main():
    print("\n=== MCP QSO Test (via Web API) ===\n")

    # Check for manual callsign specification
    if len(sys.argv) >= 3:
        target_call = sys.argv[1]
        target_grid = sys.argv[2]
        channel = sys.argv[3] if len(sys.argv) > 3 else 'C'
        instance_id = f'Slice-{channel}'

        print(f"Manual mode: Testing QSO with {target_call}")
        print(f"Target: {target_call}")
        print(f"Grid: {target_grid}")
        print(f"Channel: {channel} (20m if C)")
        print(f"Instance: {instance_id}")

        result = execute_qso(
            instance_id=instance_id,
            target_callsign=target_call,
            my_callsign='HB9BLA',
            my_grid='JN37VL'
        )

        print("\n=== QSO Started ===")
        print(json.dumps(result, indent=2))
        print("\nQSO is now running autonomously. Monitor WSJT-X window for progress.")
        return

    # Get current slice states
    print("Fetching current slice states...")
    slices = get_slices()

    # Show all channels
    print("\n=== Current Activity ===")
    for slice_data in slices:
        channel = slice_data.get('id', slice_data.get('name', '?'))
        band = slice_data.get('band', '?')
        freq = slice_data.get('dialFrequency', 0)
        stations = slice_data.get('stations', [])
        cq_count = sum(1 for s in stations if 'CQ' in s.get('message', ''))
        print(f"Channel {channel} ({band}, {freq/1e6:.3f} MHz): {len(stations)} stations, {cq_count} calling CQ")

    # Find weak CQ stations on 20m (Channel C)
    print("\nSearching for weak CQ stations on 20m (Channel C)...")
    cq_stations = find_cq_stations(slices, channel='C', min_snr=-20, max_snr=-10)

    if not cq_stations:
        print("No weak CQ stations found on 20m")
        print("\nSearching for any CQ stations on 20m...")
        cq_stations = find_cq_stations(slices, channel='C', min_snr=-30, max_snr=0)

    if not cq_stations:
        print("No CQ stations found on 20m")
        print("\nUsage: python test-qso.py <callsign> <grid> [channel]")
        print("Example: python test-qso.py PD1HPB JO22 C")
        sys.exit(1)

    # Sort by SNR (weakest first)
    cq_stations.sort(key=lambda x: x['snr'])

    print(f"\nFound {len(cq_stations)} CQ stations on 20m:")
    for i, station in enumerate(cq_stations[:5], 1):
        print(f"  {i}. {station['callsign']:10s} {station['grid']:6s} {station['snr']:+3d} dB - {station['message']}")

    # Select the weakest station
    target = cq_stations[0]

    print(f"\n=== Starting QSO with {target['callsign']} ===")
    print(f"Target: {target['callsign']}")
    print(f"Grid: {target['grid']}")
    print(f"SNR: {target['snr']:+d} dB")
    print(f"Channel: C (20m, 14.074 MHz)")
    print(f"Instance: Slice-C")

    # Execute QSO
    result = execute_qso(
        instance_id='Slice-C',
        target_callsign=target['callsign'],
        my_callsign='HB9BLA',
        my_grid='JN37VL'
    )

    print("\n=== QSO Started ===")
    print(json.dumps(result, indent=2))
    print("\nQSO is now running autonomously. Monitor WSJT-X Slice-C window for progress.")
    print("The QSO state machine will handle the complete exchange automatically.")

if __name__ == '__main__':
    try:
        main()
    except requests.exceptions.ConnectionError:
        print("Error: Cannot connect to MCP server at http://localhost:3001")
        print("Make sure the MCP server is running.")
        sys.exit(1)
    except Exception as e:
        print(f"Error: {e}")
        sys.exit(1)

File: test-qso.ts
import { Client } from '@modelcontextprotocol/sdk/client/index.js';
import { StdioClientTransport } from '@modelcontextprotocol/sdk/client/stdio.js';
import { spawn } from 'child_process';

async function testQso() {
    console.log('=== WSJT-X QSO Execution Test ===\n');

    // Create MCP client
    const transport = new StdioClientTransport({
        command: 'node',
        args: ['dist/index.js'],
    });

    const client = new Client({
        name: 'qso-test-client',
        version: '1.0.0',
    }, {
        capabilities: {},
    });

    try {
        console.log('Connecting to MCP server...');
        await client.connect(transport);
        console.log('✓ Connected to MCP server\n');

        // Step 1: Get current decodes
        console.log('Step 1: Getting current decodes...');
        const decodesResult = await client.callTool({
            name: 'wsjtx_get_decodes',
            arguments: {
                max_age_seconds: 60,
            },
        });

        if (!decodesResult.content || (decodesResult.content as any[]).length === 0) {
            console.log('✗ No decodes available');
            return;
        }

        const decodesText = (decodesResult.content as any)[0].text;
        const decodes = JSON.parse(decodesText);
        console.log(`✓ Found ${decodes.length} recent decodes\n`);

        // Step 2: Find CQ calling stations
        console.log('Step 2: Finding CQ calling stations...');
        const cqStations = decodes.filter((d: any) =>
            d.raw_text && d.raw_text.includes('CQ ')
        );

        if (cqStations.length === 0) {
            console.log('✗ No CQ calling stations found');
            console.log('Available stations:');
            decodes.slice(0, 10).forEach((d: any) => {
                console.log(`  ${d.call || 'N/A'}: ${d.raw_text} (${d.snr_db}dB)`);
            });
            return;
        }

        console.log(`✓ Found ${cqStations.length} CQ calling stations:`);
        cqStations.slice(0, 5).forEach((d: any, i: number) => {
            console.log(`  ${i + 1}. ${d.call}: ${d.raw_text} (${d.snr_db}dB @ ${d.audio_offset_hz}Hz)`);
        });

        // Step 3: Select the strongest CQ station
        const strongest = cqStations.reduce((prev: any, curr: any) =>
            (curr.snr_db > prev.snr_db) ? curr : prev
        );

        console.log(`\n✓ Selected strongest station: ${strongest.call} (${strongest.snr_db}dB)`);
        console.log(`  Message: ${strongest.raw_text}`);
        console.log(`  Frequency: ${strongest.audio_offset_hz}Hz`);
        console.log(`  Instance: ${strongest.instance_id}\n`);

        // Step 4: Get instance info to get our callsign and grid
        console.log('Step 4: Getting instance information...');
        const instancesResult = await client.callTool({
            name: 'wsjtx_list_instances',
            arguments: {},
        });

        const instancesText = (instancesResult.content as any)?.[0]?.text || '[]';
        const instances = JSON.parse(instancesText);
        const targetInstance = instances.find((i: any) => i.id === strongest.instance_id);

        if (!targetInstance) {
            console.log(`✗ Could not find instance ${strongest.instance_id}`);
            return;
        }

        console.log(`✓ Instance: ${targetInstance.id}`);
        console.log(`  My Call: ${targetInstance.myCall || 'HB9BLA'}`);
        console.log(`  My Grid: ${targetInstance.myGrid || 'JN37'}\n`);

        const myCall = targetInstance.myCall || 'HB9BLA';
        const myGrid = targetInstance.myGrid || 'JN37';

        // Step 5: Execute QSO
        console.log(`Step 5: Executing QSO with ${strongest.call}...`);
        console.log(`Command: execute_qso(${strongest.instance_id}, ${strongest.call}, ${myCall}, ${myGrid})\n`);

        const qsoResult = await client.callTool({
            name: 'execute_qso',
            arguments: {
                instance_id: strongest.instance_id,
                target_callsign: strongest.call,
                my_callsign: myCall,
                my_grid: myGrid,
            },
        });

        const qsoText = (qsoResult.content as any)?.[0]?.text || '';
        console.log('✓ QSO execution started:');
        console.log(qsoText);

        // Step 6: Monitor for a while
        console.log('\nStep 6: Monitoring QSO progress for 60 seconds...');
        console.log('(Watch WSJT-X window for TX activity)\n');

        await new Promise(resolve => setTimeout(resolve, 60000));

        console.log('\n=== Test Complete ===');

    } catch (error: any) {
        console.error('✗ Error:', error.message);
        if (error.code) {
            console.error(`  Error code: ${error.code}`);
        }
        if (error.data) {
            console.error(`  Details: ${JSON.stringify(error.data, null, 2)}`);
        }
    } finally {
        await client.close();
        console.log('Disconnected from MCP server');
    }
}

testQso().catch(console.error);
